### PR TITLE
Add NVMe diagnostics support

### PR DESCRIPTION
**Pull request by murilofv**
The NVMe diagnostics is done by using the SMART data retrieved from at an NVMe device, the data is analyzed and depending on the health status of the device it can generate events in the servicelog database, these events might be just warnings or in case of an unrecoverable error it will log a serviceable call home event for the device. Users can control which types of events will be reported through a config file that will be checked whenever the tool is executed. The NVMe diagnostics will be automatically executed on a daily basis as part of a cron script that will be automatically installed as well. At this point this tool is restricted to run in P10 PowerVM LPARs.

### CODE REVIEW

File: diags/diag_nvme.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+/* get_smart_log_page - Get SMART data for NVMe device
+ * @fd - File descriptor index of NVMe device
+ * @nsid - Namespace ID, use 0xFFFFFFFF or 0x0 to get controller data, former is preferred
+ * @log - Address of log struct to store the SMART data
+ *
+ * Do an ioctl to retrieve SMART data for the specified NSID and NVMe device
+ *
+ * Return - ioctl return code is returned, 0 on success
+ */
+int get_smart_log_page(int fd, uint32_t nsid, struct nvme_smart_log_page *log) {
```
Comment (sathvika-v) [inline diags/diag_nvme.c]: I think it is better to define these functions as static if they are not used outside of this file.

File: diags/diag_nvme.c [hunk 2]
```diff
@@ -17,8 +17,102 @@
...
+	if (tmp_rc != 0)
+		return -1;
+	snprintf(dev_path,sizeof(dev_path), "/dev/%s", device_name);
+	fd = open_nvme(dev_path);
+	if (fd < 0)
+		return -1;
+	fprintf(stdout, "Running diagnostics for %s\n", device_name);
+
+	/* After system boot VPD data might not be ready yet, so wait until VPD log page is ready */
+	while ((tmp_rc = get_ibm_vpd_log_page(fd, &vpd)) == 0x0093 && time <= MAX_TIME) {
```
Comment (sathvika-v) [inline diags/diag_nvme.c]: What does 0x0093 refer to? Can we consider using #define to define some of these values, instead of using constants?

Comment (murilofv) [reply diags/diag_nvme.c]: Sure, I can do that, I actually need to correct the value here, the proper value is supposed to be 0x00F3 to indicate that VPD data is not ready yet and we need to retry the operation later.

File: diags/diag_nvme.c (hunk 1 above)
Comment (murilofv) [reply diags/diag_nvme.c]: While not currently in use in another file, I ended up leaving the prototype of functions that I could see someone in the future wanting to reutilize to manipulate NVMe devices in the diag_nvme.h header file, while the specific functions for the diagnostics I left the prototype in the diag_nvme.c file, let me know if you think this separation is enough or if I should still label things as extern / static.

Comment (murilofv) [reply diags/diag_nvme.c]: I made the changes adding static / extern to functions similarly to what I've mentioned in my previous comment, that even though they are not currently in use elsewhere they might be useful as kind of a "common" code to manipulate NVMe devices. Let me know if you have any concerns / issues with the current approach.

File: diags/diag_nvme.c (hunk 2 above)
Comment (murilofv) [reply diags/diag_nvme.c]: Change done, thank you for the feedback.

**Review by maheshsal — COMMENTED** [review]
Thanks for fixing this.. Looks like this was lingering around quite some time.

**Comment by murilofv** [comment]
Is there any additional changes you guys identified that are needed?

**Pull request merged**